src/debug/profiler.rs

```diff
@@ -0,0 +1,183 @@
+static PROFILER_LOCK: Lazy<Mutex<()>> = Lazy::new(|| Mutex::new(()));
+
+#[derive(Clone)]
+struct UploadContext {
+    bucket: String,
+    client: Client,
+}
+
+#[derive(Deserialize)]
+pub struct ProfilerPayload {
+    profile_duration: u64,
+}
+
+pub async fn init_upload_context(config: &crate::env::Config) {
+    let mut lock = UPLOAD_CONTEXT.lock().await;
+    *lock = create_upload_context(config).await;
+
+    tracing::warn!("profiler: initialized: {}", lock.is_some());
+}
+
+async fn create_upload_context(config: &crate::env::Config) -> Option<UploadContext> {
+    if let Some(bucket) = config.analytics.export_bucket.as_deref() {
+        tracing::warn!(%bucket, "profiler: initializing profiler...");
+
+        let region_provider = RegionProviderChain::first_try(Region::new("eu-central-1"));
+        let aws_config = aws_config::from_env().region(region_provider).load().await;
+
+        let client = Client::new(&aws_config);
+
+        Some(UploadContext {
+            client,
+            bucket: bucket.to_owned(),
+        })
+    } else {
+        None
+    }
+}
+
+pub async fn handler(
+    State(state): State<Arc<AppState>>,
+    headers: HeaderMap,
+    Json(payload): Json<ProfilerPayload>,
+) -> impl IntoResponse {
+    let auth = headers.get("Authorization");
+    if auth.is_none() {
+        return (hyper::StatusCode::UNAUTHORIZED, "Unauthorized");
+    }
+
+    if let Ok(auth) = auth.unwrap().to_str() {
+        if auth != state.config.debug.secret {
+            return (hyper::StatusCode::UNAUTHORIZED, "Unauthorized");
+        }
+    }
+
+    tokio::spawn(run_profiler(payload));
+    (StatusCode::OK, "Profiler started.")
+}
+
+async fn run_profiler(payload: ProfilerPayload) {
+    let Ok(lock) = PROFILER_LOCK.try_lock() else {
+        tracing::warn!("profiler: failed to start because already running");
+        return;
+    };
+
+    if let Err(err) = profile(Duration::from_secs(payload.profile_duration)).await {
+        tracing::warn!(?err, "profiler: profiling failed");
+        return;
+    }
+
+    // Upload the profiles to S3.
+    upload_profiles().await;
+
+    drop(lock);
+}
+
+async fn profile(duration: Duration) -> anyhow::Result<()> {
+    let profiler = dhat::Profiler::new_heap();
+    tracing::warn!("profiler: begin profiling");
+
+    // Let the profiler run for specified duration.
+    tokio::time::sleep(duration).await;
+
+    drop(profiler);
+    tracing::warn!("profiler: end profiling");
+
+    Ok(())
+}
+
+async fn upload_profiles() {
+    tracing::warn!("profiler: uploading heap profiles in 30s...");
+
+    // Give some time to write profile data to disk if it's done on a different
+    // thread.
+    tokio::time::sleep(Duration::from_secs(30)).await;
+
+    let profiles_glob = "./dhat-*.json";
+
+    let Ok(file_list) = glob::glob(profiles_glob) else {
+        tracing::warn!("profiler: failed to read glob pattern");
+        return;
+    };
+
```

Comment:
Would be nice to know the error, just in-case

```suggestion
    let file_list = match glob::glob(profiles_glob) {
        Ok(file_list) => file_list,
        Err(err) => {
            tracing::warn!(?err, "profiler: failed to read glob pattern");
            return;
        }
    };

```